File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.php
class Solution {

    /**
     * @param String $s
     * @return Integer
     */
    function lengthOfLongestSubstring($s) {
        $map = [];
        $maxLength = 0;
        $left = 0;
        $n = strlen($s);

        for ($right = 0; $right < $n; $right++) {
            $char = $s[$right];

            if (isset($map[$char]) && $map[$char] >= $left) {
                $left = $map[$char] + 1;
            }
            $map[$char] = $right;
            $maxLength = max($maxLength, $right - $left + 1);
        }
        return $maxLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    left, maxLength := 0, 0
    index := make(map[byte]int)

    for right := 0; right < len(s); right++ {
        ch := s[right]
        if last, found := index[ch]; found && last >= left {
            left = last + 1
        }
        index[ch] = right
        if right - left + 1 > maxLength {
            maxLength = right - left + 1
        }
    }

    return maxLength
}